Enhance checkout docs (#270)
mirumee/saleor-docs: docs/developer/checkout.mdx

The code snippets included in this section may be run in [Playground](getting-started/graphql) or your preferred HTTP client.

> ### Why there is no cart model?

Reviewer: ```suggestion
### Why is there no cart model?
```

> Unlike most of other ecommerce platforms, Saleor has no separate model for a cart. Instead, we decited to merge it with checkout. Main reason was providing additional address context, required to calculate taxes in different countries.

Reviewer: ```suggestion
Saleor has no separate object type for shopping carts and baskets. We wanted the same features—like discounts, vouchers, address-specific taxes, and shipping estimates—to be available in the cart and the checkout, so we've decided to use the same object type for both.
```

> Checkout provides interface for usual cart operations like managing adding products or promo codes. 

Reviewer: ```suggestion
Checkout provides the interface for usual cart operations like adding products or promo codes. It can also be completed in almost any order, for example, saving a shipping address before adding any items.
```

⚠️ Please confirm the above is true.

Reviewer: That's an unusual requirement. What happens if I add a product, give my shipping address, and then remove that product? Shouldn't checkout only _require_ the shipping address if there are any shippable products and skip it otherwise, even if one was submitted?

> Checkout provides interface for usual cart operations like managing adding products or promo codes. 

Author: ```suggestion
Checkout provides the interface for usual cart operations like adding products or promo codes. It can also be completed in almost any order, for example, saving a billing address before adding any items. 
```

}
```

## Managing items

### Adding a product

[`checkoutLinesAdd`](developer/api-reference/mutations/checkout-lines-add.mdx)

```graphql
mutation {
  checkoutLinesAdd(
    checkoutId: "Q2hlY2tvdXQ6MzQwNTFhMDAtNTI3NC00OWEwLWE3YmEtMTE0NjUwMWY3MzM3"
    lines: [{ quantity: 1, variantId: "UHJvZHVjdFZhcmlhbnQ6Mjc0" }]
  ) {
    checkout {
      lines{
        variant{
          name
        }
        quantity
      }
      totalPrice {
        gross {
          currency
          amount
        }
      }
    }
  }
}
```

> Response contains updated checkout contents:

Reviewer: ```suggestion
The response contains the updated checkout object:
```

```json
{
  "data": {
    "checkoutLinesAdd": {
      "checkout": {
        "lines": [
          {
            "variant": {
              "name": "XL"
            },
            "quantity": 1
          }
        ],
        "totalPrice": {
          "gross": {
            "currency": "USD",
            "amount": 5
          }
        }
      }
    }
  }
}
```

> If variant is already added, its quantity will be increased. Running prevoius mutation twice will result in checkout with one line and quantity set to 2:

Reviewer: ```suggestion
If the variant is already added to the checkout, Saleor will increase the corresponding line's quantity. Running the previous mutation twice will result in a checkout object with one line and the quantity set to 2:
```

}
```

## Managing items

### Adding a product to checkout

> To add an item to the cart, use [`checkoutLinesAdd`](developer/api-reference/mutations/checkout-lines-add.mdx). Total prices will be updated automatically:

Reviewer 2: Would be good to include somewhere info that storefront should always fetch the prices from backend after any change in lines, shipping, or addresses as the values can be changed after applied texes.  It can be treat as something obvious but I would like to emphasize it.

Author: I'm not sure in which section it would fit now. Maybe at assigning address section?
I'll let know Zaiste that such info would be also in Saleor introduction tutorial 💪

            "variant": {
              "name": "XL"
            },
            "quantity": 2
          }
        ],
        "totalPrice": {
          "gross": {
            "currency": "USD",
            "amount": 10
          }
        }
      }
    }
  }
}
```

### Changing the line quantity

To modify line quantity, use [checkoutLinesUpdate](developer/api-reference/mutations/checkout-lines-update.mdx):

```graphql {4}
mutation {
  checkoutLinesUpdate(
    token: "4b22fbc0-b095-4937-a08e-bd18d04c951a"
    lines: [{ quantity: 12, variantId: "UHJvZHVjdFZhcmlhbnQ6MTcy" }]
  ){
    checkout{
      lines{
        id
        variant{
          id
        }
        quantity
      }
    }
    errors{
      field
      message
    }
  }
}
```

```json {11}
{
  "data": {
    "checkoutLinesUpdate": {
      "checkout": {
        "lines": [
          {
            "id": "Q2hlY2tvdXRMaW5lOjI1NA==",
            "variant": {
              "id": "UHJvZHVjdFZhcmlhbnQ6MTcy"
            },
            "quantity": 12
          }
        ]
      },
      "errors": []
    }
  }
}
```

If the quantity is changed to 0, API will remove this line from checkout.

### Removing items from checkout

> Running [checkoutLineDelete](developer/api-reference/mutations/checkout-line-delete.mdx) mutation will remove whole line, no matter the quantity:

Reviewer 3: ```suggestion
Running the [checkoutLineDelete](developer/api-reference/mutations/checkout-line-delete.mdx) mutation will remove the whole line, no matter the quantity:
```

  ){
    checkout{
      lines{
        id
        variant{
          id
        }
        quantity
      }
    }
    errors{
      field
      message
    }
  }
}
```

```json
{
  "data": {
    "checkoutLineDelete": {
      "checkout": {
        "lines": []
      },
      "errors": []
    }
  }
}
```

## Shipping
This step is only used if purchased items require shipping (if they are physical products). The user must select a specific shipping method to create shipping for this checkout. To signify whether shipping is required, use the `isShippingRequired` field in the [`Checkout`](developer/api-reference/objects/checkout.mdx) object.

<Chart
  definition={`
  sequenceDiagram
  Client ->> Saleor: checkoutCreate
  alt customer exists and has default address
    Saleor->>Saleor: Assign default shipping and billing address
  end
  Saleor -->> Client: Checkout
  alt customer chenges address
    Client ->> Saleor: new address
    Saleor -->> Client: address set
  end
  Client ->> Saleor: getShippingMethods
  Saleor -->> Client: available ShippingMethods
  Client ->> Saleor: checkoutShippingAddressUpdate
  Saleor -->> Client: updated Checkout
`}
/>

### Adding address data

> Address can be assigned using [`checkoutShippingAddressUpdate`](developer/api-reference/mutations/checkout-shipping-address-update.mdx) mutation:

Reviewer 3: ```suggestion
Address can be assigned using the [`checkoutShippingAddressUpdate`](developer/api-reference/mutations/checkout-shipping-address-update.mdx) mutation:
```

```graphql
mutation {
  checkoutShippingAddressUpdate(
    billingAddress: {
      country: PL
      firstName: "John"
      lastName: "Smith"
      streetAddress1: "ul. Tęczowa 7"
      postalCode: "53-030"
      city: "Wroclaw"
    }
    token: "58f4ca17-9c6f-437b-8204-beb47bbee364"
  ) {
    checkout {
      billingAddress {
        firstName
        lastName
        streetAddress1
        city
        postalCode
      }
    }
    errors {
      field
      message
      code
    }
  }
}
```

Successful response: 
```json
{
  "data": {
    "checkoutShippingAddressUpdate": {
      "checkout": {
        "billingAddress": {
          "firstName": "John",
          "lastName": "Smith",
          "streetAddress1": "ul. Tęczowa 7",
          "city": "WROCLAW",
          "postalCode": "53-030"
        }
      },
      "errors": []
    }
  }
}
```

> More information about address validation can be found on [address validation page](developer/address.mdx).

Reviewer 3: ```suggestion
More information about address validation can be found on the [address validation](developer/address.mdx) page.
```

}
```

## Payment

Depending on the selected payment gateway, you will either use the JavaScript form which can be integrated to Saleor, or the payment gateway will direct you to an external payment page. The payment gateway sends information if the payment is successful, along with tokenized credit card payment information. This token is then used to run the [`checkoutPaymentCreate`](developer/api-reference/mutations/checkout-payment-create.mdx) mutation.

<Chart
  definition={`
  sequenceDiagram
  Client ->>+ Saleor: getPaymentMethods
  Saleor -->>- Client: available PaymentMethods
  Client ->>+ Saleor: checkoutPaymentCreate
  Saleor -->>- Client: payment created
`}
/>

### Listing available payment gateways

> Available payment gateways can be listed from `Checkout.availablePaymentGateways` field. Depending on the chosen gateway, you may need additional configuration. For example, public `api_key` used by frontend payment library. To obtain that data, run:

Reviewer 3: ```suggestion
Available payment gateways can be listed from the `Checkout.availablePaymentGateways` field. Depending on the chosen gateway, you may need additional configuration. For example, public `api_key` is used by the frontend payment library. To obtain that data, run:
```

```graphql
query {
  checkout(token: "853ca234-b113-4021-8434-b37087d652a1"){
    availablePaymentGateways {
      id
      name
      config {
        field
        value
      }
    }
  }
}
```

Response:
```json
{
  "data": {
    "checkout": {
      "availablePaymentGateways": [
        {
          "id": "mirumee.payments.stripe",
          "name": "Stripe",
          "config": [
            {
              "field": "api_key",
              "value": "pk_test_SkWEusCNFdSgxGGmd23z7JC1FHh3uNufNuX"
            },
            {
              "field": "store_customer_card",
              "value": "false"
            }
          ]
        },
        {
          "id": "mirumee.payments.dummy",
          "name": "Dummy",
          "config": [
            {
              "field": "store_customer_card",
              "value": "false"
            }
          ]
        }
      ]
    }
  }
}
```

### Creating payment
The [`checkoutPaymentCreate`](developer/api-reference/mutations/checkout-payment-create.mdx) mutation requires

> - `token`: the checkout token

Reviewer 2: This is not true. `token` is required only for PaymentGateway which demands it (for example Braintree). For Adyen, Stripe this is not required field

## Completing the checkout

> The purpose of the operation is to ensure this checkout is correct. The validation consist of checking:

Reviewer 3: ```suggestion
The purpose of the operation is to ensure this checkout is correct. The validation consists of checking:
```
